File: src/main/java/org/wint3794/pathfollower/geometry/CurvePoint.java
package org.wint3794.pathfollower.geometry;

public class CurvePoint extends Point{
    public double moveSpeed;
    public double turnSpeed;
    public double followDistance;
    public double slowDownTurnRadians;
    public double slowDownTurnAmount;
    public double pointLength;


    public CurvePoint(double x, double y, double moveSpeed, double turnSpeed,
                      double followDistance, double slowDownTurnRadians, double slowDownTurnAmount){
        this.x = x;
        this.y = y;
        this.moveSpeed = moveSpeed;
        this.turnSpeed = turnSpeed;
        this.followDistance = followDistance;
        this.pointLength = followDistance;
        this.slowDownTurnRadians = slowDownTurnRadians;
        this.slowDownTurnAmount = slowDownTurnAmount;
    }

    public CurvePoint(Point point) {
        this.x = point.x;
        this.y = point.y;
    }


    public CurvePoint(double x, double y, double moveSpeed, double turnSpeed, double followDistance,double pointLength, double slowDownTurnRadians, double slowDownTurnAmount){

        this(x,y, moveSpeed, turnSpeed, followDistance, slowDownTurnRadians, slowDownTurnAmount);
        this.pointLength = pointLength;
    }

    public CurvePoint(CurvePoint nextPoint) {
        x = nextPoint.x;
        y = nextPoint.y;
        moveSpeed = nextPoint.moveSpeed;
        turnSpeed = nextPoint.turnSpeed;
        followDistance = nextPoint.followDistance;
        slowDownTurnRadians = nextPoint.slowDownTurnRadians;
        slowDownTurnAmount = nextPoint.slowDownTurnAmount;
        pointLength = nextPoint.pointLength;

    }

    public Point toPoint(){
        return new Point(x,y);
    }
    public void setPoint(Point p){
        x = p.x;
        y = p.y;
    }
}

File: src/main/kotlin/org/wint3794/pathfollower/geometry/CurvePoint.kt
package org.wint3794.pathfollower.geometry

/**
 * A class for mark points in virtual path
 */
data class CurvePoint(
    var x: Double = 0.0,
    var y: Double = 0.0,
    var moveSpeed: Double = 0.0,
    var turnSpeed: Double = 0.0,
    var followDistance: Double = 0.0,
    var slowDownTurnRadians: Double = 0.0,
    var slowDownTurnAmount: Double = 0.0,
    var pointLength: Double = 0.0
) {

    constructor(point: Point) {
        x = point.x
        y = point.y
    }

    constructor(nextPoint: CurvePoint) {
        x = nextPoint.x
        y = nextPoint.y
        moveSpeed = nextPoint.moveSpeed
        turnSpeed = nextPoint.turnSpeed
        followDistance = nextPoint.followDistance
        slowDownTurnRadians = nextPoint.slowDownTurnRadians
        slowDownTurnAmount = nextPoint.slowDownTurnAmount
        pointLength = nextPoint.pointLength
    }

    fun toPoint(): Point {
        return Point(x, y)
    }

    fun setPoint(p: Point) {
        x = p.x
        y = p.y
    }
}